### PR TITLE
Rework material on library components in Cabal Users Guide

### PR DESCRIPTION
**Pull request by mpilgrem**
The existing material in the Cabal Users Guide on library components is written up as incremental 'developments' on top of the 'history'. This reworking aims to recast the existing material to describe the current state of affairs.

It also aims to eliminate duplication and be consistent in the terminology used, as follows: a library component can be the 'main' unnamed library or a named library (which can be referred to as a 'sublibrary'). A sublibrary can be 'public' or 'private'. A private sublibrary can be referred to as an 'internal' library.

It moves `pkg-field:: visibility: visibility specifiers` to the top of the discussion of `library` section fields, as whether a library component is 'public' or 'private' is a fundamental characteristic of the component.

It also clears the existing `TODO inline the blog post`. I consider that all of the substance in the blog post is now to be found in the Users Guide.

### CODE REVIEW

**Review by geekosaur — APPROVED** [review]

**Comment by ffaf1** [comment]
@mpilgrem I think this and #9371 overlaps. Do you want me to close mine or do you prefer to rebase once mine is merged?

**Comment by mpilgrem** [comment]
@ffaf1, I'd be happy to rebase this once yours is merged.

**Comment by mpilgrem** [comment]
I have rebased on #9371, but also made some further changes, namely:

* I have explained the basis of the term 'sublibraries' in terms of 'subsidiary' or 'ancillary' libraries (about line 805)
* I have used plainer language than the existing text at about lines 834 to 838 (explaining the significance of 'public' and 'private')
* I have used 'expose' rather than the existing 'export' at line 919. I think modules 'export' functions etc and packages 'expose' modules.
* I haved used plainer language than the existing text at about lines 956 and 957. In particular, I did not know what 'to vendor' meant.

File: doc/cabal-package.rst
Comment (TeofilC) [inline doc/cabal-package.rst:812]:
```suggestion
    A sublibrary may be marked 'public' or 'private' (the
    default) — see the :pkg-field:`library:visibility` field. The main
    library is always public. A private sublibrary is also known as
    an internal library. 
    Only public library components can be depended on by other packages.
```
I'm suggesting:
  - only quoting things when necessary
  - changing a dash to an em-dash
  - dropping the reference to "unnamed". I don't think that's relevant here
  - split the last sentence into two.

File: doc/cabal-package.rst
Comment (TeofilC) [inline doc/cabal-package.rst:820]:
```suggestion
    Being able to include more than one public library in a package allows separating the
    unit of distribution (the package) from the unit of buildable code (the
    library). This is useful for Haskell projects with many libraries that are
    distributed together as it avoids duplication and potential inconsistencies.
```
This reads slightly better to me but I feel like this is a matter of taste

File: doc/cabal-package.rst
Comment (TeofilC) [inline doc/cabal-package.rst:832]:
```suggestion
        ``private`` for sublibraries. Cannot be set for the main
        library, which is always public.
```
As above, I don't think "main" needs to be quoted here and unnamed doesn't add any extra detail.

File: doc/cabal-package.rst
Comment (TeofilC) [inline doc/cabal-package.rst:838]:
```suggestion
    Can be set to ``public`` or ``private``. If set to ``public``, then any component can depend on this library.
    If set to ``private``, then only components from this package can depend on it.
```

File: doc/cabal-package.rst
Comment (TeofilC) [inline doc/cabal-package.rst:841]:
```suggestion
    See the :pkg-field:`build-depends` field for the syntax to specify a dependency on a main library or a sublibrary.
```
As above I don't think we need to quote and called it unnamed. Also let's not say "build-depends" twice, rather let's say what it represents.

File: doc/cabal-package.rst
Comment (TeofilC) [inline doc/cabal-package.rst:919]:
```suggestion
is a test suite that needs access to some internal modules in your main
library, which you do not otherwise want to expose. You could put those
```
As above

File: doc/cabal-package.rst
Comment (TeofilC) [inline doc/cabal-package.rst:917]:
```suggestion
An example of the use of an internal library (that is, a private sublibrary)
```
I don't think this needs to be quoted

File: doc/cabal-package.rst
Comment (TeofilC) [inline doc/cabal-package.rst:953]: I feel like this cuts quite a bit. Some of this overlaps with the 'visibility' field, but it's good to have this info here as well I think. Especially the explanation of shadowing seems important.

File: doc/cabal-package.rst
Comment (TeofilC) [inline doc/cabal-package.rst:956]: Again the use of shadowing is key to this example. I don't think it should be cut

File: doc/cabal-package.rst
Comment (TeofilC) [inline doc/cabal-package.rst:1439]:
```suggestion
    Declares the *library* dependencies required to build the current package component. See
```
I don't think this parenthesis is adding any extra information. You are saying either X or not X, so that's everything

File: doc/cabal-package.rst
Comment (TeofilC) [inline doc/cabal-package.rst:1447]:
```suggestion
    A library is identified by the name of its package,
    optionally followed by a colon and the library's name (available from
```

File: doc/cabal-package.rst
Comment (TeofilC) [inline doc/cabal-package.rst:1449]:
```suggestion
    ``cabal-version: 3.0``). If a library name is omitted, the package's main
    library will be used. To refer expressly to a package's 'main'
```
As above.

File: doc/cabal-package.rst
Comment (TeofilC) [inline doc/cabal-package.rst:1456]:
```suggestion
    See :pkg-section:`library` for more information about declaring
    library components.
```
Be explicit about what this is about: declaring library components.

I think some of this is unnecessary and can be cut

**Review by TeofilC — COMMENTED** [review]
Hey, I had a read through this and had some ideas for improving this PR.
I'm not a maintainer so feel free to take my suggestions with a pinch of salt.

It might seem like a lot of comments but a lot of the suggestions are the same thing repeated a couple of times.

If you have the chance to take a look at one thing then it's having another look at the internal library example. I think some information is lost in your edits

**Comment by TeofilC** [comment]
> I haved used plainer language than the existing text at about lines 956 and 957. In particular, I did not know what 'to vendor' meant.

"to vendor" normally means to include some external source code, eg, if you are writing bindings to a C library you can vendor that library by including the C source as part of your source tarball. In the example we are vendoring parts of attoparsec by including their source in our library rather than depending on attoparsec from Hackage, eg, because it would lead to a cyclic dependency

**Comment by ulysses4ever** [comment]
I don't thing this use of the word "vendor" is universally recognized. So, I think, it should be either explained or replaced with something (macro-expanded with your explanation, perhaps).

**Comment by geekosaur** [comment]
It's pretty widely recognized in the developer community.

**Comment by ffaf1** [comment]
I was unsure of its meaning before this conversation, had to check
Wiktionary.

    https://en.wiktionary.org/wiki/vendor#Verb

I find “to bundle” is more understandable to me as a non-native English
speaker. Is this an appropriate synonym?


Il 28 ottobre 2023 alle 11:55 brandon s allbery kf8nh ha scritto:

**Comment by geekosaur** [comment]
Perhaps both (`bundle (vendor)`), since to me "bundle" does not imply use, just inclusion.

File: doc/cabal-package.rst
Comment (mpilgrem) [reply doc/cabal-package.rst:13]: I agree. My proposed change (not leading with the ancient history, but not deleting it) pulled its punches.

File: doc/cabal-package.rst
Comment (mpilgrem) [reply doc/cabal-package.rst:807]: I'm happy to split into shorter sentences. My reason for referring to 'subsidiary' and 'ancillary' is to explain the (likely) origin of (and so give content to) the abbreviated term 'sublibrary' for named library components. I included both because some people preferred 'ancilliary' to 'subsidiary'. (Their meanings are very similar.)

Comment (geekosaur) [reply doc/cabal-package.rst:807]: I'm currently wondering if "related library" is better and clearer than either. (I was originally in favor of "ancillary", but…
```
[26 01:45:36] <haskellbridge> <g​eekosaur> I considered "ancillary" but half expect people to look at me like I said "monad"…
```
(The defense rests.))

Comment (TeofilC) [reply doc/cabal-package.rst:807]: Haha very prescient!

File: doc/cabal-package.rst
Comment (mpilgrem) [reply doc/cabal-package.rst:832]: As above, this was my attenpt to avoid chosing sides in the named/unamed v main/sub debate.

**Comment by mpilgrem** [comment]
On the use of terms such as 'to vendor', I don't know what audience the Cabal Users Guide has in mind. When I edit the Stack Users Guide, I try to have in mind somebody leaving school and about to start a course on computer science at university; it helps me with the 'assumed prior knowledge of the reader'. I substituted 'to incorporate' for 'to vendor'.

File: doc/cabal-package.rst
Comment (mpilgrem) [reply doc/cabal-package.rst:841]: I agree.

File: doc/cabal-package.rst
Comment (mpilgrem) [reply doc/cabal-package.rst:953]: On shadowing, I thought it no longer applies (from Cabal 3.4), because a sublibrary is now specified in a `build-depends:` as `package_name:library_name`. I think shaddowing was only applicable when a sublibrary could be specified as just `library_name` (which might clash with a package's name).

On the rest that is cut, it is all directly duplicative.

File: doc/cabal-package.rst
Comment (mpilgrem) [reply doc/cabal-package.rst:956]: As above, I don't think shaddowing applies, from Cabal 3.4.

File: doc/cabal-package.rst
Comment (mpilgrem) [reply doc/cabal-package.rst:1439]: I think it is useful to be express that a component of a package can depend on other (library) components in the same package AND library components in other packages (i.e. on 'everything').

(The fact that an executable can depend on the library in the same package is the 'Cabal 1.8.0.4' feature that was highlighted in the opening sentences of this part of the guide.)

File: doc/cabal-package.rst
Comment (mpilgrem) [reply doc/cabal-package.rst:812]: There has been debate about whether library components should be referred to as (a) unnamed/named or (b) main/sub[sidiary]. The former is objective, but does not convey the relative 'standing' of the named library component. The latter conveys the (subjective) relative 'standing'. I avoided that choice by using "'main' unnamed library".

The quotation marks around the other words on their first introduction were also because they were being used figuratively rather than literally. An alternative would be to use code font (eg ``visibility``).

File: doc/cabal-package.rst
Comment (mpilgrem) [reply doc/cabal-package.rst:820]: I agree.

**Comment by mpilgrem** [comment]
I'll wait to see what views other people have before attempting to assimilate them all in a revised proposal.

File: doc/cabal-package.rst
Comment (mpilgrem) [reply doc/cabal-package.rst:953]: On Cabal 3.4, see: https://cabal.readthedocs.io/en/stable/file-format-changelog.html#cabal-version-3-4:
> Dependencies to sublibraries must be specified explicitly, even for current package. For example: build-depends: mypackage:internal-lib This way you can have an internal library with the same name as some external dependency.

File: doc/cabal-package.rst
Comment (geekosaur) [reply doc/cabal-package.rst:956]: This is talking about the file format, not the Cabal version. Note also that currently Hackage refuses versions later than 3.0 on the grounds that they may not be generally usable yet (although I think this is under review), so I'd consider 3.0 the breakpoint for talking about cabal file format versions.

Comment (mpilgrem) [reply doc/cabal-package.rst:956]: My reasoning was that the body of the Guide should discuss the current version of the library and the tool, and the corresponding version of the Cabal specification - because if the Guide was to discuss (at any length) the history of the specification each time it discussed a current feature of the library/tool its complexity could be off-putting to readers and, perhaps, more difficult to maintain. However, if Hackage is de facto limiting the adoption of versions of the specification, I agree it is useful for the Guide to deal with relevant history of the specification back to the point that Hackage accepts. I'll make edits to reflect to the relevant history of the specification of sublibraries in `build-depends:`, and reintroduce 'shadowing' - but not as a 'feature', given that it has subsequently been 'fixed'.

**Comment by mpilgrem** [comment]
On reflection, I have force pushed an alternative now. @TeofilC's comments have encourged me to be more muscular in rephrasing existing content. I have now:

* removed the pre-Cabal 1.8.0.4 history from the 'Package Description' introduction (about lines 11 to 13);
* dealt with the named/unnamed v main/sub[sidiary] debate by first introducing the structure in objective terms (lines 800 to 804) and then introducing the shorthand that the guide subsequently uses consistently (lines 811 to 814);
* included the fact that cabal-install does not, currently, support packages with sublibraries but no main library (at lines 806 to 809);
* deal with the history of the Cabal specification as regards shadowing at lines 818 to 822 and 1471 to 1476;
* use 'Notes' to cover the history of the Cabal specification before the current specification;
* reintroduced, parenthetically, 'vendor' and 'bundle', at line 973; and
* been consistent with the form of the examples provided at lines 1463 to 1469.

File: doc/cabal-package.rst
Comment (mpilgrem) [reply doc/cabal-package.rst]: I have now taken a different approach to this, by changing the introduction of terms.

File: doc/cabal-package.rst
Comment (mpilgrem) [reply doc/cabal-package.rst]: I have accepted this, as the point I wanted to make express is now expressed elsewhere.

File: doc/cabal-package.rst
Comment (mpilgrem) [reply doc/cabal-package.rst]: This is now reflected.

File: doc/cabal-package.rst
Comment (mpilgrem) [reply doc/cabal-package.rst]: This is now reflected. The history of the Cabal specification has been moved out of the paragraph into a note.

File: doc/cabal-package.rst
Comment (mpilgrem) [reply doc/cabal-package.rst]: I have now shortened this. I preferred 'to specify' over 'to declare', because the former seems more frequent in the existing documentation. I don't think much turns on that (although editorial consistency would be nice to have).

File: doc/cabal-package.rst
Comment (mpilgrem) [reply doc/cabal-package.rst]: Reflected in my subsequent changes.

File: doc/cabal-package.rst
Comment (mpilgrem) [reply doc/cabal-package.rst]: Reflected in my subsequent changes.

File: doc/cabal-package.rst
Comment (mpilgrem) [reply doc/cabal-package.rst:970]: 'Shadowing' brought back in, elsewhere.

**Comment by mpilgrem** [comment]
Rebased on `master`.

**Comment by andreabedini** [comment]
You all are doing an amazing job here! ❤️

**Comment by mpilgrem** [comment]
Rebased on `master`.

**Comment by TeofilC** [comment]
Thanks for the ping! I'll take another look either Tuesday or Wednesday

**Review by TeofilC — APPROVED** [review]
I read over the latest version and it looks great to me! Thanks for all your work on this!

**Comment by ulysses4ever** [comment]
I elevated permissions for @TeofilC to Write so that his review counts towards the threshold.

**Pull request merged**
